src/AdvancedMH.jl

```diff
@@ -0,0 +1,135 @@
+module AdvancedMH
+
+# Import the relevant libraries.
+using Reexport
+using Turing.Interface
+@reexport using Distributions
+using Random
+
+# Import specific functions and types to use or overload.
+import Distributions: VariateForm, ValueSupport, variate_form, value_support, Sampleable
+import MCMCChains: Chains
+import Turing.Interface: step!, AbstractSampler, AbstractTransition, transition_type
+
+# Exports
+export MetropolisHastings, DensityModel, sample
+
+# Define a sampler type.
+struct MetropolisHastings{T} <: AbstractSampler 
+    init_θ :: T
+    proposal :: Function
+end
+
+# Default constructors.
+MetropolisHastings(init_θ::T) where T<:Real = MetropolisHastings{T}(init_θ, x -> Normal(x, 1.0))
+MetropolisHastings(init_θ::T) where T<:Vector{<:Real} = MetropolisHastings{T}(init_θ, x -> MvNormal(x, 1.0))
+
+# Define a model type. Stores the log density function and the data to 
+# evaluate the log density on.
+struct DensityModel{V<:VariateForm, S<:ValueSupport, T} <: Sampleable{V, S}
+    π :: Function
+    data :: T
+end
+
+# Default density constructor.
+DensityModel(π::Function, data::T) where T = DensityModel{VariateForm, ValueSupport, T}(π, data)
```

Comment:
Why do we need this?

Reply:
We need this to pass in the abstract types `VariateForm` and `ValueSupport`, because otherwise users would have to annoyingly specify it themselves. Most people are unlikely to care, so this is just a convenience function. This is part of allowing models to be subtypes of `Sampleable`, which we haven't really though too hard about -- it's kind of left in the interface just in case we want to step the modelling language up to the next level.